### DISCUSSION TITLE
Flask-SocketIO reconnects Websocket after HTTP request

### DISCUSSION THREAD

**Discussion by BillyHesse** [miguelgrinberg/Flask-SocketIO#1822] [Q&A]
# **Flask-SocketIO reconnects Websocket after HTTP request**

Hey Python Webdev's!

I'm currently working on a Flask-Webserver (with Flask-Login, Flask-SocketIO and Flask-Session) to control relais for static and also flashing lights. Using uWSGI with Gevent-Websocket for deployment i am encountering the problem, that the User who triggers an HTTP request gets disconnected from the Websocket. After 4-8 seconds the Websocket reconnects the User with a new SID and everything works again. While all other Users (anonymous and also authenticated) just recieve the data from the socket instantly as supposed to. I already studied logs and tried everything to solve this problem in the past 10 days and couldn't find any error.

**Comment by miguelgrinberg** [comment]
Is the hardware code that you are using compatible with gevent? Most hardware control libraries aren't.

**Reply by BillyHesse** [reply]
Thanks for the fast reply @miguelgrinberg! It was working when i used werkzeug development server with simple-websocket. Just after the deployment version the problem accured.

I am using a raspberry pi with RPi.GPIO

**Reply by miguelgrinberg** [reply]
Sure, but now you switched from Werkzeug to Gevent. You can only use code that is compatible with gevent.

**Reply by BillyHesse** [reply]
Right after i updated from flask-socketio 4.X to 5.X the problem accured. 
Lately i've uninstalled uWSGI, gevent etc and ran it again with socketio .run() it also didn't work. I made sure it would ran on the same settings as in 4.X

I'll try on a werkzeug server again

**Reply by BillyHesse** [reply]
Switching back to the werkzeug server with simple-websocket also didn't work.

Could the emty response from return ("", 204) be a part of the problem?

The live updates work on all other clients (or users), just the one who clicked a button gets reconnected. It's suspicious the problems accured after updating all modules. 
(RPi.GPIO wasn't updated)

**Reply by miguelgrinberg** [reply]
Okay, I think we may be confusing two separate issues.

Using the GPIO library with gevent is a problem. Maybe not your main problem, but definitely a problem.

I haven't seen the code that runs when you click buttons, so I can't comment on that, but I don't see how that can be a problem with this library. The only way a button click can interrupt a connection is if you allow the browser to reload the page. If you are doing that, then this is a design flaw in your application, which should be written as a single page app to ensure that the WebSocket connections do not break.

**Reply by BillyHesse** [reply]
I've looked deeply into SocketIO.run() and the Flask.run() method and found out, maybe the working development-server version was using wsgi with eventlet and greenlet. To be honest, i didn't note down any of the choices SocketIO.run() made, but i can remember reading "Werkzeug runinng on". Just to make shure this isn't a dejavue i'll give the wsgi server with green(threads) a go.

My main guess is, that the standard threading with gevent could cause this problem.

The function that gets executed on a button click is shown above: update_relais_mode(RX)
I really, really made shure there is only non-blocking code.

**Reply by miguelgrinberg** [reply]
Gevent and eventlet are both incompatible with GPIO, unless your library has been specifically designed for async use.

You have avoided my question though. What happens when the user clicks the button? Does the browser reload the page? That's where the problem likely is, not in the server.

**Reply by BillyHesse** [reply]
When clicking a button the browser doesn't reload. I've tried MS Edge, Chrome and Samsung Browser.

I thought "return ("", 204)" doesn't trigger any reloading

**Reply by miguelgrinberg** [reply]
What are you doing to prevent the browser from reloading? I think this is likely a misunderstanding on your part regarding how the browser works.

**Reply by BillyHesse** [reply]
## You're right seems like the page is reloading.

I think i understood your reply now. You may be right with the reloading-issue because i am using JS with window.location.href to follow the btn link.

## At this point the HTML side does this (for every button)
### body
img src="/static/btn_rd_light_active.png" id="btn_rd_light_active" class="btn_rd_light_active">

### JS
rd_light = function(event) {
      window.location.href = "/btnchg/btn_rd_light";}

### From Stackoverflow
The internet suggests not using JS to make the HTTP request. Instead i should use button> or input type="button"/>#
https://stackoverflow.com/questions/1878264/how-do-i-make-an-html-button-not-reload-the-page
